torch/jit/__init__.py

```diff
@@ -580,5 +570,77 @@ def _time(trace_name, name, time=True):
         print("{} {} time: {} ms".format(trace_name, name, start.elapsed_time(end)))
 
 
+def verify(model, args, loss_fn=torch.sum, devices=None):
+    """
+    Verify that a JIT compiled model has the same behavior as its uncompiled
+    version along with its backwards pass.
+
+    This function has side-effects (e.g., it executes your model / saves and loads
+    parameters), so don't expect the model to come out exactly the same as what
+    you passed in.
+
+    Arguments:
+        model (compiled torch.nn.Module or function): the module/function to be
+            verified.  The module/function definition MUST have been decorated with
+            `@torch.jit.compile`.
+        args (tuple or Variable): the positional arguments to pass to the
+            compiled function/module to be verified.  A non-tuple is assumed to
+            be a single positional argument to be passed to the model.
+        loss_fn (function, optional): the loss function to be applied to
+            the output of the model, before backwards is invoked.  By default,
+            we assume that a model returns a single result, and we :func:`torch.sum`
+            before calling backwards; if this is inappropriate, you can pass your
+            own loss function.  Note that if a model returns a tuple of results,
+            these are passed as separate arguments to `loss_fn`.
+        devices (iterable of device IDs, optional): the GPU devices which the
+            compiled module will be run on.  This determines the RNG state we
+            must save when running both compiled and uncompiled versions of the model.
+    """
+    # TODO: In principle, we track device information in our trace, so it
+    # should be possible to check if our execution actually obeyed the 'devices'
+    # the user provided.
+
+    # TODO: Consider adding a utility function to torch.jit to test
+    # for this case
+    if not isinstance(model, _CompiledMixin):
+        raise TypeError("Cannot verify an uncompiled module.  Add @torch.jit.compile to compile it")
+
+    if not isinstance(args, tuple):
+        args = (args,)
+
+    saved_args = _clone_inputs(args)
+    saved_state = copy.deepcopy(model.state_dict())
+
+    def run_fwd_bwd(args, force_trace=False):
+        in_vars, _ = _flatten(args, model.state_dict(keep_vars=True).values())
+        # We use a special API to reset the trace and compile it from scratch.
+        out = model(*args, _force_trace=force_trace)
+        if not isinstance(out, tuple):
+            out = (out, )
+        out_vars, _ = _flatten(out)
+        saved_outs = [v.data.clone() for v in out_vars]
+        loss = loss_fn(*out)
```

Comment:
You can't assume to do `*out` in general. If model outputs a tuple, you won't wrap it later, and you will unpack it in here.

Reply:
Actually, this was intentional: *if* the model returns a tuple, I want the loss function to handle it as a multivariadic function; if it returns just a single output, you just get one argument. It's more pleasant to write lambdas this way, because you can't destructure tuples in arguments.